### ISSUE TITLE
app.py

### ISSUE THREAD

**Issue by vijayakumar-cmd** [vijayakumar-cmd/AI-ANC#1]
"""
app.py — SENTINEL-ANC single-file Hugging Face Space
=====================================================
Everything needed to serve the DTLN speech enhancer lives in this one file:
the model definition, checkpoint loading, and the Gradio UI. No imports from
models/ or utils/, so you can drop this file alone into a Space.

FILES YOU NEED IN THE SPACE
---------------------------
    app.py            <- this file
    requirements.txt  <- 5 lines, see README section at the bottom
    best.pth          <- YOUR trained checkpoint (upload after training)

Without best.pth the app still starts, but it runs with random weights and
says so loudly in the UI. That is useful for testing the deployment before
your training finishes.
"""

from __future__ import annotations

import os
from pathlib import Path
from typing import Tuple

import gradio as gr
import numpy as np
import torch
import torch.nn as nn
import torch.nn.functional as F

SAMPLE_RATE = 16_000
CKPT_PATH = os.environ.get("CKPT", "best.pth")

# ============================================================================ #
# MODEL — identical to models/dtln_model.py, inlined so this file stands alone
# ============================================================================ #
class InstantLayerNorm(nn.Module):
    """Per-frame layer norm. Causal: no statistics leak across time frames."""

    def __init__(self, num_features: int, eps: float = 1e-7) -> None:
        super().__init__()
        self.eps = eps
        self.gamma = nn.Parameter(torch.ones(1, 1, num_features))
        self.beta = nn.Parameter(torch.zeros(1, 1, num_features))

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        mean = torch.mean(x, dim=-1, keepdim=True)
        var = torch.mean((x - mean) ** 2, dim=-1, keepdim=True)
        return (x - mean) / torch.sqrt(var + self.eps) * self.gamma + self.beta

class SeparationCore(nn.Module):
    """Stacked causal LSTM + dense + sigmoid -> multiplicative mask in (0, 1)."""

    def __init__(self, input_size: int, hidden_size: int, output_size: int,
                 num_layers: int = 2, dropout: float = 0.0) -> None:
        super().__init__()
        self.lstm = nn.LSTM(input_size, hidden_size, num_layers,
                            batch_first=True, bidirectional=False,
                            dropout=dropout if num_layers > 1 else 0.0)
        self.dense = nn.Linear(hidden_size, output_size)

    def forward(self, x, state=None):
        out, new_state = self.lstm(x, state)
        return torch.sigmoid(self.dense(out)), new_state

class DTLN(nn.Module):
    """Dual-Signal Transformation LSTM Network. ~989k parameters at defaults."""

    def __init__(self, block_len: int = 512, block_shift: int = 128,
                 hidden_size: int = 128, encoder_size: int = 256,
                 num_layers: int = 2, dropout: float = 0.0) -> None:
        super().__init__()
        self.block_len = block_len
        self.block_shift = block_shift
        self.freq_bins = block_len // 2 + 1

        self.core_1 = SeparationCore(self.freq_bins, hidden_size, self.freq_bins,
                                     num_layers, dropout)
        self.encoder = nn.Conv1d(block_len, encoder_size, 1, bias=False)
        self.encoder_norm = InstantLayerNorm(encoder_size)
        self.core_2 = SeparationCore(encoder_size, hidden_size, encoder_size,
                                     num_layers, dropout)
        self.decoder = nn.Conv1d(encoder_size, block_len, 1, bias=False)

    def _frame(self, x: torch.Tensor) -> torch.Tensor:
        if x.size(-1) < self.block_len:
            x = F.pad(x, (0, self.block_len - x.size(-1)))
        return x.unfold(-1, self.block_len, self.block_shift)

    def _overlap_add(self, frames: torch.Tensor, length: int) -> torch.Tensor:
        batch = frames.size(0)
        out = F.fold(frames.transpose(1, 2), output_size=(1, length),
                     kernel_size=(1, self.block_len),
                     stride=(1, self.block_shift)).reshape(batch, length)
        return out * (self.block_shift / self.block_len)

    def forward(self, waveform: torch.Tensor) -> torch.Tensor:
        squeeze_back = waveform.dim() == 1
        if squeeze_back:
            waveform = waveform.unsqueeze(0)

        orig_len = waveform.size(-1)
        n_frames = max(1, 1 + (orig_len - self.block_len + self.block_shift - 1)
                       // self.block_shift)
        padded_len = (n_frames - 1) * self.block_shift + self.block_len
        frames = self._frame(F.pad(waveform, (0, padded_len - orig_len)))

        # --- Core 1: mask the STFT magnitude, keep the noisy phase ---
        spec = torch.fft.rfft(frames, n=self.block_len, dim=-1)
        mask_1, _ = self.core_1(torch.abs(spec))
        est_spec = torch.polar(torch.abs(spec) * mask_1, torch.angle(spec))
        frames_1 = torch.fft.irfft(est_spec, n=self.block_len, dim=-1)

        # --- Core 2: mask in a learned basis to repair phase artefacts ---
        encoded = self.encoder(frames_1.transpose(1, 2)).transpose(1, 2)
        mask_2, _ = self.core_2(self.encoder_norm(encoded))
        frames_2 = self.decoder((encoded * mask_2).transpose(1, 2)).transpose(1, 2)

        enhanced = self._overlap_add(frames_2, padded_len)[:, :orig_len]
        return enhanced.squeeze(0) if squeeze_back else enhanced

    def count_parameters(self) -> int:
        return sum(p.numel() for p in self.parameters())

def si_sdr(estimate: torch.Tensor, target: torch.Tensor, eps: float = 1e-8):
    """Scale-invariant SDR in dB, per example."""
    estimate = estimate - estimate.mean(-1, keepdim=True)
    target = target - target.mean(-1, keepdim=True)
    proj = (estimate * target).sum(-1, keepdim=True) / \
           ((target ** 2).sum(-1, keepdim=True) + eps) * target
    noise = estimate - proj
    return 10 * torch.log10(((proj ** 2).sum(-1) + eps) / ((noise ** 2).sum(-1) + eps))

# ============================================================================ #
# LOAD ONCE AT STARTUP
# ============================================================================ #
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
ckpt_file = Path(CKPT_PATH)
HAS_WEIGHTS = ckpt_file.is_file()

if HAS_WEIGHTS:
    ckpt = torch.load(ckpt_file, map_location="cpu")
    cfg = ckpt.get("config", {})
    model = DTLN(block_len=cfg.get("block_len", 512),
                 block_shift=cfg.get("block_shift", 128),
                 hidden_size=cfg.get("hidden_size", 128),
                 encoder_size=cfg.get("encoder_size", 256),
                 dropout=0.0)
    # strict=False tolerates minor key drift between training and demo code
    missing, unexpected = model.load_state_dict(ckpt["model_state"], strict=False)
    if missing or unexpected:
        print(f"State dict warning — missing: {missing}, unexpected: {unexpected}")
    epoch = ckpt.get("epoch", "?")
    best = ckpt.get("best_si_sdr", float("nan"))
    STATUS = f"Trained weights loaded — epoch {epoch}, best val SI-SDR {best:.2f} dB"
    BANNER = ""
else:
    model = DTLN()
    STATUS = f"NO CHECKPOINT at '{CKPT_PATH}' — running with RANDOM weights."
    BANNER = (
        "### Demo mode: untrained\n"
        f"No `{CKPT_PATH}` found, so the network has random weights and the output "
        "will be noise. The deployment itself is working — upload your trained "
        "`best.pth` to this Space and it will switch over automatically on restart."
    )

model.to(device).eval()
print(STATUS, f"| params: {model.count_parameters():,} | device: {device}")

# ============================================================================ #
# INFERENCE
# ============================================================================ #
@torch.no_grad()
def enhance(audio: Tuple[int, np.ndarray] | None):
    if audio is None:
        return None, "Upload a file or record from the microphone first."

    sr, data = audio
    raw = np.asarray(data)
    x = raw.astype(np.float32)

    if x.ndim > 1:                                    # stereo -> mono
        x = x.mean(axis=1)
    if np.issubdtype(raw.dtype, np.integer):          # int16 -> float [-1, 1]
        x = x / float(np.iinfo(raw.dtype).max)

    peak_in = float(np.abs(x).max())
    if peak_in < 1e-6:
        return None, "That clip is silent."

    wav = torch.from_numpy(x).unsqueeze(0)

    if sr != SAMPLE_RATE:                             # resample to 16 kHz
        try:
            import torchaudio
            wav = torchaudio.functional.resample(wav, sr, SAMPLE_RATE)
        except ImportError:
            n = int(wav.shape[-1] * SAMPLE_RATE / sr)
            wav = F.interpolate(wav.unsqueeze(1), size=n,
                                mode="linear", align_corners=False).squeeze(1)

    duration = wav.shape[-1] / SAMPLE_RATE
    if duration > 60:
        return None, f"Clip is {duration:.0f}s. Please keep it under 60s."

    # Pad one block each side: overlap-add is attenuated at the signal edges.
    pad = model.block_len
    wav_p = F.pad(wav, (pad, pad))

    try:
        enhanced = model(wav_p.to(device)).cpu()[:, pad:-pad]
    except Exception as exc:
        return None, f"Inference failed: {exc}"

    peak = enhanced.abs().max()
    if peak > 1.0:
        enhanced = enhanced / peak

    retained = float(si_sdr(enhanced, wav).mean())
    report = (
        f"{STATUS}\n"
        f"Parameters : {model.count_parameters():,}\n"
        f"Device     : {device}\n"
        f"Input      : {duration:.2f}s, resampled {sr} -> {SAMPLE_RATE} Hz\n"
        f"SI-SDR vs input: {retained:.2f} dB  "
        f"(how much of the original mixture survived, not a quality score)"
    )
    return (SAMPLE_RATE, enhanced.squeeze(0).numpy()), report

# ============================================================================ #
# UI
# ============================================================================ #
with gr.Blocks(title="SENTINEL-ANC") as demo:
    gr.Markdown(
        "# SENTINEL-ANC — DTLN Speech Enhancement\n"
        "Dual-Signal Transformation LSTM Network, ~989k parameters, 16 kHz mono. "
        "Cloud prototype of a real-time noise suppressor targeting NVIDIA Jetson."
    )
    if BANNER:
        gr.Markdown(BANNER)

    with gr.Row():
        inp = gr.Audio(sources=["upload", "microphone"], type="numpy",
                       label="Noisy input")
        out = gr.Audio(label="Enhanced output")
    btn = gr.Button("Enhance", variant="primary")
    rep = gr.Textbox(label="Run report", lines=6)

    btn.click(fn=enhance, inputs=inp, outputs=[out, rep])
    inp.change(fn=enhance, inputs=inp, outputs=[out, rep])

if __name__ == "__main__":
    demo.launch(server_name="0.0.0.0",
                server_port=int(os.environ.get("PORT", 7860)))
